extensions/pkg/controller/controlplane/genericactuator/actuator.go

```diff
@@ -525,3 +506,69 @@ func (a *actuator) Migrate(
 ) error {
 	return a.Delete(ctx, cp, cluster)
 }
+
+// ReconcileShootWebhooks deploys the shoot webhook configuration, i.e., a network policy to allow the kube-apiserver to
+// talk to the provider extension, and a managed resource that contains the MutatingWebhookConfiguration.
+func ReconcileShootWebhooks(ctx context.Context, c client.Client, namespace, providerName string, serverPort int, shootWebhooks []admissionregistrationv1beta1.MutatingWebhook) error {
+	if err := extensionswebhookshoot.EnsureNetworkPolicy(ctx, c, namespace, providerName, serverPort); err != nil {
+		return errors.Wrapf(err, "could not create or update network policy for shoot webhooks in namespace '%s'", namespace)
+	}
+
+	webhookConfiguration, err := marshalWebhooks(shootWebhooks, providerName)
+	if err != nil {
+		return err
+	}
+
+	if err := manager.
+		NewSecret(c).
+		WithNamespacedName(namespace, ShootWebhooksResourceName).
+		WithKeyValues(map[string][]byte{"mutatingwebhookconfiguration.yaml": webhookConfiguration}).
+		Reconcile(ctx); err != nil {
+		return errors.Wrapf(err, "could not create or update secret '%s/%s' of managed resource containing shoot webhooks", namespace, ShootWebhooksResourceName)
+	}
+
+	if err := manager.
+		NewManagedResource(c).
+		WithNamespacedName(namespace, ShootWebhooksResourceName).
+		WithSecretRef(ShootWebhooksResourceName).
+		Reconcile(ctx); err != nil {
+		return errors.Wrapf(err, "could not create or update managed resource '%s/%s' containing shoot webhooks", namespace, ShootWebhooksResourceName)
+	}
+
+	return nil
+}
+
+// ReconcileShootWebhooksForAllNamespaces reconciles the shoot webhooks in all shoot namespaces of the given
+// provider type. This is necessary in case the webhook port is changed (otherwise, the network policy would only be
+// updated again as part of the ControlPlane reconciliation which might only happen in the next 24h).
+func ReconcileShootWebhooksForAllNamespaces(ctx context.Context, c client.Client, providerName, providerType string, port int, shootWebhooks []admissionregistrationv1beta1.MutatingWebhook) error {
+	namespaceList := &corev1.NamespaceList{}
+	if err := c.List(ctx, namespaceList, client.MatchingLabels{
+		v1beta1constants.GardenRole:         v1beta1constants.GardenRoleShoot,
+		v1beta1constants.LabelShootProvider: providerType,
+	}); err != nil {
+		return err
+	}
+
+	fns := make([]flow.TaskFn, 0, len(namespaceList.Items))
+
+	for _, namespace := range namespaceList.Items {
+		var (
+			networkPolicy     = extensionswebhookshoot.GetNetworkPolicyMeta(namespace.Name, providerName)
+			namespaceName     = namespace.Name
+			networkPolicyName = networkPolicy.Name
+		)
+
+		fns = append(fns, func(ctx context.Context) error {
+			if err := c.Get(ctx, kutil.Key(namespaceName, networkPolicyName), &networkingv1.NetworkPolicy{}); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return err
+				}
+				return nil
+			}
```

Comment:
This code skips the webhook reconciliation if the networkpolicy was not found. Why is this needed / when would this happen? Isn't it better to reconcile anyway, just in case?

Reply:
For example, when a shoot is already in the process of being deleted and the webhook configuration has already been cleaned up (then we don't want to recreate it unnecessarily).